drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_rtio_stream.c

```diff
@@ -0,0 +1,344 @@
+	lsm6dsv16x_fifo_temp_batch_t temp_batch = LSM6DSV16X_DT_TEMP_NOT_BATCHED;
+	lsm6dsv16x_fifo_mode_t fifo_mode = LSM6DSV16X_BYPASS_MODE;
+
+	pin_int.fifo_th = PROPERTY_DISABLE;
+
+	if (enable != 0) {
+		pin_int.fifo_th = PROPERTY_ENABLE;
+
+		xl_batch = config->accel_batch;
+		gy_batch = config->gyro_batch;
+		temp_batch = config->temp_batch;
+
+		fifo_mode = LSM6DSV16X_STREAM_MODE;
+		fifo_wtm = config->fifo_wtm;
+	}
+
+	/*
+	 * Set FIFO watermark (number of unread sensor data TAG + 6 bytes
+	 * stored in FIFO) to FIFO_WATERMARK samples
+	 */
+	lsm6dsv16x_fifo_watermark_set(ctx, config->fifo_wtm);
+
+	/* Turn on/off FIFO */
+	lsm6dsv16x_fifo_mode_set(ctx, fifo_mode);
+
+	/* Set FIFO batch rates */
+	lsm6dsv16x_fifo_xl_batch_set(ctx, xl_batch);
+	lsm6dsv16x->accel_batch_odr = xl_batch;
+	lsm6dsv16x_fifo_gy_batch_set(ctx, gy_batch);
+	lsm6dsv16x->gyro_batch_odr = gy_batch;
+#if defined(CONFIG_LSM6DSV16X_ENABLE_TEMP)
+	lsm6dsv16x_fifo_temp_batch_set(ctx, temp_batch);
+	lsm6dsv16x->temp_batch_odr = temp_batch;
+#endif
+
+	/* Set pin interrupt (fifo_th could be on or off) */
+	if (config->drdy_pin == 1) {
+		lsm6dsv16x_pin_int1_route_set(ctx, &pin_int);
+	}  else {
+		lsm6dsv16x_pin_int2_route_set(ctx, &pin_int);
+	}
+}
+
+void lsm6dsv16x_submit_stream(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	struct lsm6dsv16x_data *lsm6dsv16x = dev->data;
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	uint8_t fifo_irq = 0;
+
+	gpio_pin_interrupt_configure_dt(lsm6dsv16x->drdy_gpio, GPIO_INT_DISABLE);
+
+	for (size_t i = 0; i < cfg->count; i++) {
+		if ((cfg->triggers[i].trigger == SENSOR_TRIG_FIFO_WATERMARK) ||
+		    (cfg->triggers[i].trigger == SENSOR_TRIG_FIFO_FULL)) {
+			fifo_irq = 1;
+		}
+	}
+
+	/* if any change in fifo irq */
+	if (fifo_irq != lsm6dsv16x->fifo_irq) {
+		lsm6dsv16x->fifo_irq = fifo_irq;
+
+		/* enable/disable the FIFO */
+		lsm6dsv16x_config_fifo(dev, fifo_irq);
+	}
+
+	lsm6dsv16x->streaming_sqe = iodev_sqe;
+
+	gpio_pin_interrupt_configure_dt(lsm6dsv16x->drdy_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+static void lsm6dsv16x_complete_op_cb(struct rtio *r, const struct rtio_sqe *sqe, void *arg)
+{
+	const struct device *dev = arg;
+	struct lsm6dsv16x_data *lsm6dsv16x = dev->data;
+
+	/*
+	 * Mark operation completed
+	 */
+	rtio_iodev_sqe_ok(sqe->userdata, 0);
+	lsm6dsv16x->streaming_sqe = NULL;
+	gpio_pin_interrupt_configure_dt(lsm6dsv16x->drdy_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+}
+
+static void lsm6dsv16x_read_fifo_cb(struct rtio *r, const struct rtio_sqe *sqe, void *arg)
+{
+	const struct device *dev = arg;
+	struct lsm6dsv16x_data *lsm6dsv16x = dev->data;
+	struct gpio_dt_spec *irq_gpio = lsm6dsv16x->drdy_gpio;
+	struct rtio_iodev *iodev = lsm6dsv16x->iodev;
+	struct sensor_read_config *read_config;
+	uint8_t fifo_th = 0, fifo_full = 0;
+	uint16_t fifo_count;
+
+	/* proceed only if streaming */
+	if (lsm6dsv16x->streaming_sqe == NULL) {
```

Comment:
This seems like it'd be a programming mistake at this point and should perhaps be an assert

Reply:
Seems legit: if there is no request queued at this point (was already checked before that is not the case) it must be considered a bug.